<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.8.1" />
<title>test API documentation</title>
<meta name="description" content="" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>test</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">import os, sys
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from Bio import SeqIO
import itertools


class Proteome(object):
    &#34;&#34;&#34; Some useful calculations on proteomes.&#34;&#34;&#34;
    def __init__(self):
        self.type = None
        
        
    def load_uniprot(self, uniprot):
        &#34;&#34;&#34; 
        Read a UniProt proteome file.
        
        Note that the input proteome file should contain the following column order:
        (1) Entry (2) Entry name (3) Status (4) Protein names (5) Gene names (6) Organism (7) Length (8) Sequence 
            
        Parameters
        ----------              
        `uniprot` : file downloaded from UniProt containing a proteome of interest

        
        Returns
        -------
        `proteome_dataframe`, a Pandas dataframe that contains the FASTA IDs, sequences, and length of sequences 
        &#34;&#34;&#34;        
        # Load the dataset with column headers defined
        self.uniprot_proteome = pd.read_csv(uniprot, sep=&#39;\t&#39;, names=[&#39;Entry&#39;, &#39;Entry name&#39;, &#39;Status&#39;, &#39;Protein names&#39;, &#39;Gene names&#39;, &#39;Organism&#39;, &#39;Length&#39;, &#39;Sequence&#39;])
        
        # Convert protein length values to integers
        self.uniprot_proteome[&#39;Length&#39;][1:] =  self.uniprot_proteome[&#39;Length&#39;][1:].astype(int)  
        
        return self.uniprot_proteome
        
 
    def load_fasta(self, fasta):
        &#34;&#34;&#34; 
        Read a FASTA file with identifiers and sequences.
        
        Note that the FASTA file can be single- or multi-line.
            
        Parameters
        ----------              
        `fasta` : file in FASTA format with identifiers (e.g. &gt;) and sequences 
        
        Returns
        -------
        `fasta_proteome`, a Pandas dataframe that contains the FASTA IDs, sequences, and length of sequences
        &#34;&#34;&#34;
        # Load the fasta file
        self.seq_list = SeqIO.index(fasta, &#34;fasta&#34;)  # makes dictionary that contains Seq objects

        # Create arrays to store ID and sequence information
        ids = []  ; seqs = []
        for key, value in self.seq_list.items():
            ids.append(key)
            seqs.append(str(value.seq)) # convert the Seq object to a string

        # Create a dictionary for subsequent conversion into a dataframe
        self.dictionary = {&#39;ID&#39;:ids, &#39;Sequence&#39;:seqs} 

        # Load the dictionary into a dataframe
        self.fasta_proteome = pd.DataFrame(self.dictionary)  
        self.fasta_proteome[&#39;Length&#39;] = self.fasta_proteome[&#39;Sequence&#39;].str.len()  # create column in the dataframe with sequence lengths
        self.fasta_proteome[&#39;Length&#39;][1:].astype(int)   # convert the length values to integers

        return self.fasta_proteome
        
        
    def proteome_length(self, query_proteome):
        &#34;&#34;&#34; 
        Calculate the total number of amino acids in a proteome.
            
        Parameters
        ----------              
        `query_proteome` : Pandas dataframe, contains the proteome of interest
        
        Returns
        -------
        `total_length` :  int, total number of residues in proteome_dataframe
        &#34;&#34;&#34;   
        # Sum up the total number of residues
        total_length = query_proteome[&#39;Length&#39;][1:].sum(axis=0, skipna=True)
        
        return total_length
        
        
    def proteome_amino_acid_fractions(self, query_proteome):
        &#34;&#34;&#34; 
        Calculate the fractional amino acid composition of a proteome.
            
        Parameters
        ----------              
        `query_proteome` : Pandas dataframe, contains the proteome of interest
        
        Returns
        -------
        `AA_fractions` :  Pandas dataframe, the fraction of each amino acid
        &#34;&#34;&#34;   
        # Amino acids 
        self.amino_acids = [&#39;A&#39;, &#39;C&#39;, &#39;D&#39;, &#39;E&#39;, &#39;F&#39;, &#39;G&#39;, &#39;H&#39;, &#39;I&#39;, &#39;K&#39;, &#39;L&#39;, &#39;M&#39;, &#39;N&#39;, &#39;P&#39;, &#39;Q&#39;, &#39;R&#39;, &#39;S&#39;, &#39;T&#39;, &#39;V&#39;, &#39;W&#39;, &#39;Y&#39;]
        
        # Initialize an empty array to store the fractional amino acid values
        fractions = []
    
        # Compute the fractional amino acid composition of the proteome
        for aa in self.amino_acids:
            fractions.append(self.find_motifs(query_proteome, aa)[1])
            
        # Convert into a dictionary and then a Pandas dataframe 
        AA_fractions = pd.DataFrame(list(dict(zip(self.amino_acids, fractions)).items()), columns=[&#39;AA&#39;, &#39;Fraction&#39;])
        AA_fractions[&#39;Fraction&#39;][1:].astype(float) # convert the fractions to floats
        
        return AA_fractions  
 
 
    def find_motifs(self, data_frame, query):
        &#34;&#34;&#34; 
        Search a proteome for a specific amino acid or motif.
            
        Parameters
        ----------              
        `data_frame` : Pandas dataframe, contains the proteome of interest
        
        `query` :  str, the amino acid or motif to be searched
        
        Returns
        -------
        `sum` :  int, total number of hits for the queried residue/motif
        
        `fraction`:  float, sum divided by the total number of residues or motifs of the same size
        &#34;&#34;&#34;  
        # Extract sequences from the pandas dataframe
        seqs = data_frame[&#39;Sequence&#39;]
                
        # Calculate the total number of residues in the queried proteome
        total_residues = self.proteome_length(data_frame)
        
        # Initialize an array
        hits_array = []
        
        # Loop over each sequence and count the number of times the queried motif shows up
        for protein in seqs:
            hits_array.append(protein.count(query))

        # total number of hits for the queried residue/motif, check if any hits are found
        sum = np.sum(np.asarray(hits_array))  
        if sum &lt; 1:
            fraction = 0
        else:
            fraction = sum/(total_residues - len(query)+1)  # total number of hits divided by the total number of residues or motifs of the same size
                        
        return sum, fraction       
 
        
    def count_multiple_motifs(self, data_frame, list_of_motifs):
        &#34;&#34;&#34; 
        Search a proteome for a list of motifs.
            
        Parameters
        ----------              
        `data_frame` : int, contains the proteome of interest
        
        `list_of_motifs` :  list, the amino acids or motifs to be searched.
        
        Returns
        -------
        `counted_motifs` :  Pandas dataframe
            Lists the queried motifs and values.
        &#34;&#34;&#34;         
        # Initialize arrays
        query_array = []  ; found_array = []  ; length_array = []
        
        # Loop over the list of motifs and count the number of instances
        for i, motif in enumerate(list_of_motifs):
            query_array.append(motif)
            found_array.append(self.find_motifs(data_frame, motif)[0])
            length_array.append(len(motif))
        
        # Create an array with the total number of possible motifs for each motif
        all_motifs = np.zeros(shape=(len(length_array),))
        for x in range(0, len(length_array)):
            all_motifs[x] = self.proteome_length(data_frame) - length_array[x]

        # Convert into a dictionary and then a Pandas dataframe 
        counted_motifs = pd.DataFrame(list(dict(zip(query_array, found_array)).items()), columns=[&#39;Motif&#39;, &#39;Count&#39;])
        counted_motifs[&#39;Count&#39;][1:].astype(int)  # make the Count column integer format         
        counted_motifs[&#39;Fraction&#39;] = np.asarray(found_array)/np.asarray(all_motifs)  # add a column Fraction for the observed # motifs divided by the total # motifs
    
        return counted_motifs

        
    def iterate_motif(self, query):
        &#34;&#34;&#34; 
        Generate a list of all possible motifs that are consistent with an input motif 
        
        Note the following usage specifications:
        
        (1) the &#34;_&#34; symbol separates positions in the motif
        
        (2) &#34;X&#34; designates any residue
        
        (3) multiple residues indicate &#34;or&#34; (e.g. IV_X means Ile or Val in the 1st position followed by any residue)
        
        Example:  MILV_X_DE means Met, Ile, Leu, or Val in the 1st position followed by any residue in the 2nd position followed by Asp or Glu in the 3rd position.
        
        Parameters
        ----------              
        `query` : str, the amino acid motif to be queried 
                use &#34;_&#34; to separate positions in the motif, e.g. I/V-X-I/V would be IV_X_IV
        
        Returns
        -------
        `all_possibilities` :  list, all variations of the inputted motif (e.g. IV_X_IV yields [IAI, ICI, IDI, ... , VVV, VWV, VYV]
        &#34;&#34;&#34; 
        # Split the input query into its corresponding residue positions
        listq = [item for item in query.split(&#34;_&#34;)]

        # Generator an iterable list of motifs  
        positions = [[] for q in listq]
        for i in range(len(listq)):
            if len(listq[i]) == 1:
                if listq[i] == &#34;X&#34;:
                    for aa in self.amino_acids:
                        positions[i].append(aa)
                else:
                    positions[i].append(listq[i])
            elif len(listq[i]) &gt; 1:
                for j in listq[i]:
                    positions[i].append(j)
            else:
                print(&#34;WARNING! iterate_motif does not recognize your input&#34;)
                print(&#39;exiting now....&#39;)
                sys.exit()
        
        # Use itertools to generate the final list 
        all_possibilities = [&#39;&#39;.join(x) for x in list(itertools.product(*positions))]

        return all_possibilities
        
    def expected_motifs(self, query_proteome, query_motif):
        &#34;&#34;&#34; 
        Calculate the expected number of a queried motif based on amino acid frequencies in the proteome.
            
        Parameters
        ----------              
        `query_proteome` : Pandas dataframe
            Contains the proteome of interest.
        
        `query_motif` : str or list
            Contains the motif(s) of interest.
        
        Returns
        -------
        `final_dataframe` :  Pandas dataframe
            Expected fraction and count of a given motif. 
        &#34;&#34;&#34;   
        # Get fractional values of the amino acids within the proteome
        fractional_AA = self.proteome_amino_acid_fractions(query_proteome)
        
        # Get length of the queried proteome 
        length = self.proteome_length(query_proteome)
        
        # Create a dictionary with amino acid-to-number mapping
        numbers = {a: b for b, a in dict(enumerate(self.amino_acids)).items()}
        
        # Check if query_motif is a list of strings
        if isinstance(query_motif, list) == True:
            probabilities = np.zeros(shape=(len(query_motif)))  # Create a numpy array to store probabilities for each motif
            expected = np.zeros(shape=(len(query_motif)))  # Create a numpy array to store expected total for each motif
            observed = np.zeros(shape=(len(query_motif)))  # Create a numpy array to store observed total for each motif
            for i, motif in enumerate(query_motif):
                for res_num in range(0, len(motif)):
                    if res_num == 0:
                        probabilities[i] = fractional_AA.at[numbers[motif[res_num]], &#39;Fraction&#39;]  # get the value at the specified position
                    else:
                        probabilities[i] *= fractional_AA.at[numbers[motif[res_num]], &#39;Fraction&#39;] # multiply by the value at the specified position
                expected[i] = int(probabilities[i] * (length - len(motif)-1))  # multiply the probability and the total number of motifs
                observed[i] = int(self.find_motifs(query_proteome, motif)[0])
                
            # Combine into final Pandas dataframe    
            final_dataframe = pd.DataFrame({&#39;Motif&#39;:query_motif, &#39;Probability&#39;:probabilities, &#39;Observed&#39;:observed, &#39;Expected&#39;:expected})
        
        
        # Check if query_motif is a string
        elif isinstance(query_motif, str) == True:
            probabilities = np.zeros(shape=(1,))  # Create a numpy array to store probabilities for each motif
            expected = np.zeros(shape=(1,))  # Create a numpy array to store expected total for each motif
            observed = np.zeros(shape=(1,))  # Create a numpy array to store observed total for each motif
            for res_num in range(0, len(query_motif)):
                if res_num == 0:
                    probabilities[0] = fractional_AA.at[numbers[query_motif[res_num]], &#39;Fraction&#39;]  # get the value at the specified position
                else:
                    probabilities[0] *= fractional_AA.at[numbers[query_motif[res_num]], &#39;Fraction&#39;] # multiply by the value at the specified position
                expected[0] = int(probabilities[0] * (length - len(query_motif)-1))  # multiply the probability and the total number of motifs
                observed[0] = int(self.find_motifs(query_proteome, query_motif)[0])
                
            # Combine into final Pandas dataframe    
            final_dataframe = pd.DataFrame({&#39;Motif&#39;:query_motif, &#39;Probability&#39;:probabilities, &#39;Observed&#39;:observed, &#39;Expected&#39;:expected})
            
        final_dataframe[&#39;Observed&#39;][1:].astype(int)  # make the observed column integer format 
        final_dataframe[&#39;Expected&#39;][1:].astype(int)  # make the expected column integer format
        
        return final_dataframe
        
        
    def find_proteins(self, data_frame, query, proteome_type):
        &#34;&#34;&#34; 
        Search for proteins in a proteome that contain the queried motif.
            
        Parameters
        ----------              
        `data_frame` : Pandas dataframe, contains the proteome of interest
        
        `query` :  str or list of strings, the amino acid or motifs to be searched (e.g. &#39;IPV&#39; or [&#39;IPI&#39;, &#39;IPV&#39;, &#39;VPI&#39;, &#39;VPV&#39;])
        
        `proteome_type` : str, either &#39;FASTA&#39; or &#39;UniProt&#39; depending on the analyzed proteome
        
        Returns
        -------
        `found_proteins` :  list, array of UniProt/FASTA IDs that contain the queried residue/motif
        &#34;&#34;&#34;  
        # First check which type of proteome/Pandas dataframe is being queried
        if proteome_type.upper() == &#39;FASTA&#39;:
            id = &#39;ID&#39;
        elif proteome_type.upper() == &#39;UNIPROT&#39;:
            id = &#39;Entry&#39;
        else:
            print(&#39;ERROR! The proteome_type %s is not recognized in the function &#34;find_proteins&#34; !&#39; % proteome_type)
            print(&#39;Please enter &#34;fasta&#34; or &#34;uniprot&#34;, or modify the function to accommodate other types&#39;)
            print(&#39;Exiting now.........................................................................&#39;)
            sys.exit()

        # Initialize an empty array
        found_proteins = []
        
        # Check if queried motif (i.e. query) is a list
        if isinstance(query, list) == True:
            for i in range(0, len(data_frame[&#39;Sequence&#39;])):
                for j, motif in enumerate(query):
                    if motif in data_frame[&#39;Sequence&#39;][i]:
                        found_proteins.append(data_frame[id][i])
              
        # If not list, check if query is a string
        elif isinstance(query, str) == True:
            for i in range(0, len(data_frame[&#39;Sequence&#39;])):
                if query in data_frame[&#39;Sequence&#39;][i]:
                    found_proteins.append(data_frame[id][i])

        else:
            print(&#39;In the function &#34;find_proteins&#34;, the &#34;query&#34; must be either a list of strings or a single string....!&#39;)
            print(&#39;Please check what you entered!&#39;)
            print(&#39;Exiting now...................&#39;)
            sys.exit()
        
        # Remove duplicates from list
        found_proteins = list(dict.fromkeys(found_proteins))
        
        return found_proteins</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="test.Proteome"><code class="flex name class">
<span>class <span class="ident">Proteome</span></span>
</code></dt>
<dd>
<div class="desc"><p>Some useful calculations on proteomes.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class Proteome(object):
    &#34;&#34;&#34; Some useful calculations on proteomes.&#34;&#34;&#34;
    def __init__(self):
        self.type = None
        
        
    def load_uniprot(self, uniprot):
        &#34;&#34;&#34; 
        Read a UniProt proteome file.
        
        Note that the input proteome file should contain the following column order:
        (1) Entry (2) Entry name (3) Status (4) Protein names (5) Gene names (6) Organism (7) Length (8) Sequence 
            
        Parameters
        ----------              
        `uniprot` : file downloaded from UniProt containing a proteome of interest

        
        Returns
        -------
        `proteome_dataframe`, a Pandas dataframe that contains the FASTA IDs, sequences, and length of sequences 
        &#34;&#34;&#34;        
        # Load the dataset with column headers defined
        self.uniprot_proteome = pd.read_csv(uniprot, sep=&#39;\t&#39;, names=[&#39;Entry&#39;, &#39;Entry name&#39;, &#39;Status&#39;, &#39;Protein names&#39;, &#39;Gene names&#39;, &#39;Organism&#39;, &#39;Length&#39;, &#39;Sequence&#39;])
        
        # Convert protein length values to integers
        self.uniprot_proteome[&#39;Length&#39;][1:] =  self.uniprot_proteome[&#39;Length&#39;][1:].astype(int)  
        
        return self.uniprot_proteome
        
 
    def load_fasta(self, fasta):
        &#34;&#34;&#34; 
        Read a FASTA file with identifiers and sequences.
        
        Note that the FASTA file can be single- or multi-line.
            
        Parameters
        ----------              
        `fasta` : file in FASTA format with identifiers (e.g. &gt;) and sequences 
        
        Returns
        -------
        `fasta_proteome`, a Pandas dataframe that contains the FASTA IDs, sequences, and length of sequences
        &#34;&#34;&#34;
        # Load the fasta file
        self.seq_list = SeqIO.index(fasta, &#34;fasta&#34;)  # makes dictionary that contains Seq objects

        # Create arrays to store ID and sequence information
        ids = []  ; seqs = []
        for key, value in self.seq_list.items():
            ids.append(key)
            seqs.append(str(value.seq)) # convert the Seq object to a string

        # Create a dictionary for subsequent conversion into a dataframe
        self.dictionary = {&#39;ID&#39;:ids, &#39;Sequence&#39;:seqs} 

        # Load the dictionary into a dataframe
        self.fasta_proteome = pd.DataFrame(self.dictionary)  
        self.fasta_proteome[&#39;Length&#39;] = self.fasta_proteome[&#39;Sequence&#39;].str.len()  # create column in the dataframe with sequence lengths
        self.fasta_proteome[&#39;Length&#39;][1:].astype(int)   # convert the length values to integers

        return self.fasta_proteome
        
        
    def proteome_length(self, query_proteome):
        &#34;&#34;&#34; 
        Calculate the total number of amino acids in a proteome.
            
        Parameters
        ----------              
        `query_proteome` : Pandas dataframe, contains the proteome of interest
        
        Returns
        -------
        `total_length` :  int, total number of residues in proteome_dataframe
        &#34;&#34;&#34;   
        # Sum up the total number of residues
        total_length = query_proteome[&#39;Length&#39;][1:].sum(axis=0, skipna=True)
        
        return total_length
        
        
    def proteome_amino_acid_fractions(self, query_proteome):
        &#34;&#34;&#34; 
        Calculate the fractional amino acid composition of a proteome.
            
        Parameters
        ----------              
        `query_proteome` : Pandas dataframe, contains the proteome of interest
        
        Returns
        -------
        `AA_fractions` :  Pandas dataframe, the fraction of each amino acid
        &#34;&#34;&#34;   
        # Amino acids 
        self.amino_acids = [&#39;A&#39;, &#39;C&#39;, &#39;D&#39;, &#39;E&#39;, &#39;F&#39;, &#39;G&#39;, &#39;H&#39;, &#39;I&#39;, &#39;K&#39;, &#39;L&#39;, &#39;M&#39;, &#39;N&#39;, &#39;P&#39;, &#39;Q&#39;, &#39;R&#39;, &#39;S&#39;, &#39;T&#39;, &#39;V&#39;, &#39;W&#39;, &#39;Y&#39;]
        
        # Initialize an empty array to store the fractional amino acid values
        fractions = []
    
        # Compute the fractional amino acid composition of the proteome
        for aa in self.amino_acids:
            fractions.append(self.find_motifs(query_proteome, aa)[1])
            
        # Convert into a dictionary and then a Pandas dataframe 
        AA_fractions = pd.DataFrame(list(dict(zip(self.amino_acids, fractions)).items()), columns=[&#39;AA&#39;, &#39;Fraction&#39;])
        AA_fractions[&#39;Fraction&#39;][1:].astype(float) # convert the fractions to floats
        
        return AA_fractions  
 
 
    def find_motifs(self, data_frame, query):
        &#34;&#34;&#34; 
        Search a proteome for a specific amino acid or motif.
            
        Parameters
        ----------              
        `data_frame` : Pandas dataframe, contains the proteome of interest
        
        `query` :  str, the amino acid or motif to be searched
        
        Returns
        -------
        `sum` :  int, total number of hits for the queried residue/motif
        
        `fraction`:  float, sum divided by the total number of residues or motifs of the same size
        &#34;&#34;&#34;  
        # Extract sequences from the pandas dataframe
        seqs = data_frame[&#39;Sequence&#39;]
                
        # Calculate the total number of residues in the queried proteome
        total_residues = self.proteome_length(data_frame)
        
        # Initialize an array
        hits_array = []
        
        # Loop over each sequence and count the number of times the queried motif shows up
        for protein in seqs:
            hits_array.append(protein.count(query))

        # total number of hits for the queried residue/motif, check if any hits are found
        sum = np.sum(np.asarray(hits_array))  
        if sum &lt; 1:
            fraction = 0
        else:
            fraction = sum/(total_residues - len(query)+1)  # total number of hits divided by the total number of residues or motifs of the same size
                        
        return sum, fraction       
 
        
    def count_multiple_motifs(self, data_frame, list_of_motifs):
        &#34;&#34;&#34; 
        Search a proteome for a list of motifs.
            
        Parameters
        ----------              
        `data_frame` : int, contains the proteome of interest
        
        `list_of_motifs` :  list, the amino acids or motifs to be searched.
        
        Returns
        -------
        `counted_motifs` :  Pandas dataframe
            Lists the queried motifs and values.
        &#34;&#34;&#34;         
        # Initialize arrays
        query_array = []  ; found_array = []  ; length_array = []
        
        # Loop over the list of motifs and count the number of instances
        for i, motif in enumerate(list_of_motifs):
            query_array.append(motif)
            found_array.append(self.find_motifs(data_frame, motif)[0])
            length_array.append(len(motif))
        
        # Create an array with the total number of possible motifs for each motif
        all_motifs = np.zeros(shape=(len(length_array),))
        for x in range(0, len(length_array)):
            all_motifs[x] = self.proteome_length(data_frame) - length_array[x]

        # Convert into a dictionary and then a Pandas dataframe 
        counted_motifs = pd.DataFrame(list(dict(zip(query_array, found_array)).items()), columns=[&#39;Motif&#39;, &#39;Count&#39;])
        counted_motifs[&#39;Count&#39;][1:].astype(int)  # make the Count column integer format         
        counted_motifs[&#39;Fraction&#39;] = np.asarray(found_array)/np.asarray(all_motifs)  # add a column Fraction for the observed # motifs divided by the total # motifs
    
        return counted_motifs

        
    def iterate_motif(self, query):
        &#34;&#34;&#34; 
        Generate a list of all possible motifs that are consistent with an input motif 
        
        Note the following usage specifications:
        
        (1) the &#34;_&#34; symbol separates positions in the motif
        
        (2) &#34;X&#34; designates any residue
        
        (3) multiple residues indicate &#34;or&#34; (e.g. IV_X means Ile or Val in the 1st position followed by any residue)
        
        Example:  MILV_X_DE means Met, Ile, Leu, or Val in the 1st position followed by any residue in the 2nd position followed by Asp or Glu in the 3rd position.
        
        Parameters
        ----------              
        `query` : str, the amino acid motif to be queried 
                use &#34;_&#34; to separate positions in the motif, e.g. I/V-X-I/V would be IV_X_IV
        
        Returns
        -------
        `all_possibilities` :  list, all variations of the inputted motif (e.g. IV_X_IV yields [IAI, ICI, IDI, ... , VVV, VWV, VYV]
        &#34;&#34;&#34; 
        # Split the input query into its corresponding residue positions
        listq = [item for item in query.split(&#34;_&#34;)]

        # Generator an iterable list of motifs  
        positions = [[] for q in listq]
        for i in range(len(listq)):
            if len(listq[i]) == 1:
                if listq[i] == &#34;X&#34;:
                    for aa in self.amino_acids:
                        positions[i].append(aa)
                else:
                    positions[i].append(listq[i])
            elif len(listq[i]) &gt; 1:
                for j in listq[i]:
                    positions[i].append(j)
            else:
                print(&#34;WARNING! iterate_motif does not recognize your input&#34;)
                print(&#39;exiting now....&#39;)
                sys.exit()
        
        # Use itertools to generate the final list 
        all_possibilities = [&#39;&#39;.join(x) for x in list(itertools.product(*positions))]

        return all_possibilities
        
    def expected_motifs(self, query_proteome, query_motif):
        &#34;&#34;&#34; 
        Calculate the expected number of a queried motif based on amino acid frequencies in the proteome.
            
        Parameters
        ----------              
        `query_proteome` : Pandas dataframe
            Contains the proteome of interest.
        
        `query_motif` : str or list
            Contains the motif(s) of interest.
        
        Returns
        -------
        `final_dataframe` :  Pandas dataframe
            Expected fraction and count of a given motif. 
        &#34;&#34;&#34;   
        # Get fractional values of the amino acids within the proteome
        fractional_AA = self.proteome_amino_acid_fractions(query_proteome)
        
        # Get length of the queried proteome 
        length = self.proteome_length(query_proteome)
        
        # Create a dictionary with amino acid-to-number mapping
        numbers = {a: b for b, a in dict(enumerate(self.amino_acids)).items()}
        
        # Check if query_motif is a list of strings
        if isinstance(query_motif, list) == True:
            probabilities = np.zeros(shape=(len(query_motif)))  # Create a numpy array to store probabilities for each motif
            expected = np.zeros(shape=(len(query_motif)))  # Create a numpy array to store expected total for each motif
            observed = np.zeros(shape=(len(query_motif)))  # Create a numpy array to store observed total for each motif
            for i, motif in enumerate(query_motif):
                for res_num in range(0, len(motif)):
                    if res_num == 0:
                        probabilities[i] = fractional_AA.at[numbers[motif[res_num]], &#39;Fraction&#39;]  # get the value at the specified position
                    else:
                        probabilities[i] *= fractional_AA.at[numbers[motif[res_num]], &#39;Fraction&#39;] # multiply by the value at the specified position
                expected[i] = int(probabilities[i] * (length - len(motif)-1))  # multiply the probability and the total number of motifs
                observed[i] = int(self.find_motifs(query_proteome, motif)[0])
                
            # Combine into final Pandas dataframe    
            final_dataframe = pd.DataFrame({&#39;Motif&#39;:query_motif, &#39;Probability&#39;:probabilities, &#39;Observed&#39;:observed, &#39;Expected&#39;:expected})
        
        
        # Check if query_motif is a string
        elif isinstance(query_motif, str) == True:
            probabilities = np.zeros(shape=(1,))  # Create a numpy array to store probabilities for each motif
            expected = np.zeros(shape=(1,))  # Create a numpy array to store expected total for each motif
            observed = np.zeros(shape=(1,))  # Create a numpy array to store observed total for each motif
            for res_num in range(0, len(query_motif)):
                if res_num == 0:
                    probabilities[0] = fractional_AA.at[numbers[query_motif[res_num]], &#39;Fraction&#39;]  # get the value at the specified position
                else:
                    probabilities[0] *= fractional_AA.at[numbers[query_motif[res_num]], &#39;Fraction&#39;] # multiply by the value at the specified position
                expected[0] = int(probabilities[0] * (length - len(query_motif)-1))  # multiply the probability and the total number of motifs
                observed[0] = int(self.find_motifs(query_proteome, query_motif)[0])
                
            # Combine into final Pandas dataframe    
            final_dataframe = pd.DataFrame({&#39;Motif&#39;:query_motif, &#39;Probability&#39;:probabilities, &#39;Observed&#39;:observed, &#39;Expected&#39;:expected})
            
        final_dataframe[&#39;Observed&#39;][1:].astype(int)  # make the observed column integer format 
        final_dataframe[&#39;Expected&#39;][1:].astype(int)  # make the expected column integer format
        
        return final_dataframe
        
        
    def find_proteins(self, data_frame, query, proteome_type):
        &#34;&#34;&#34; 
        Search for proteins in a proteome that contain the queried motif.
            
        Parameters
        ----------              
        `data_frame` : Pandas dataframe, contains the proteome of interest
        
        `query` :  str or list of strings, the amino acid or motifs to be searched (e.g. &#39;IPV&#39; or [&#39;IPI&#39;, &#39;IPV&#39;, &#39;VPI&#39;, &#39;VPV&#39;])
        
        `proteome_type` : str, either &#39;FASTA&#39; or &#39;UniProt&#39; depending on the analyzed proteome
        
        Returns
        -------
        `found_proteins` :  list, array of UniProt/FASTA IDs that contain the queried residue/motif
        &#34;&#34;&#34;  
        # First check which type of proteome/Pandas dataframe is being queried
        if proteome_type.upper() == &#39;FASTA&#39;:
            id = &#39;ID&#39;
        elif proteome_type.upper() == &#39;UNIPROT&#39;:
            id = &#39;Entry&#39;
        else:
            print(&#39;ERROR! The proteome_type %s is not recognized in the function &#34;find_proteins&#34; !&#39; % proteome_type)
            print(&#39;Please enter &#34;fasta&#34; or &#34;uniprot&#34;, or modify the function to accommodate other types&#39;)
            print(&#39;Exiting now.........................................................................&#39;)
            sys.exit()

        # Initialize an empty array
        found_proteins = []
        
        # Check if queried motif (i.e. query) is a list
        if isinstance(query, list) == True:
            for i in range(0, len(data_frame[&#39;Sequence&#39;])):
                for j, motif in enumerate(query):
                    if motif in data_frame[&#39;Sequence&#39;][i]:
                        found_proteins.append(data_frame[id][i])
              
        # If not list, check if query is a string
        elif isinstance(query, str) == True:
            for i in range(0, len(data_frame[&#39;Sequence&#39;])):
                if query in data_frame[&#39;Sequence&#39;][i]:
                    found_proteins.append(data_frame[id][i])

        else:
            print(&#39;In the function &#34;find_proteins&#34;, the &#34;query&#34; must be either a list of strings or a single string....!&#39;)
            print(&#39;Please check what you entered!&#39;)
            print(&#39;Exiting now...................&#39;)
            sys.exit()
        
        # Remove duplicates from list
        found_proteins = list(dict.fromkeys(found_proteins))
        
        return found_proteins</code></pre>
</details>
<h3>Methods</h3>
<dl>
<dt id="test.Proteome.count_multiple_motifs"><code class="name flex">
<span>def <span class="ident">count_multiple_motifs</span></span>(<span>self, data_frame, list_of_motifs)</span>
</code></dt>
<dd>
<div class="desc"><p>Search a proteome for a list of motifs.</p>
<h2 id="parameters">Parameters</h2>
<p><code>data_frame</code> : int, contains the proteome of interest</p>
<p><code>list_of_motifs</code> :
list, the amino acids or motifs to be searched.</p>
<h2 id="returns">Returns</h2>
<p><code>counted_motifs</code> :
Pandas dataframe
Lists the queried motifs and values.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def count_multiple_motifs(self, data_frame, list_of_motifs):
    &#34;&#34;&#34; 
    Search a proteome for a list of motifs.
        
    Parameters
    ----------              
    `data_frame` : int, contains the proteome of interest
    
    `list_of_motifs` :  list, the amino acids or motifs to be searched.
    
    Returns
    -------
    `counted_motifs` :  Pandas dataframe
        Lists the queried motifs and values.
    &#34;&#34;&#34;         
    # Initialize arrays
    query_array = []  ; found_array = []  ; length_array = []
    
    # Loop over the list of motifs and count the number of instances
    for i, motif in enumerate(list_of_motifs):
        query_array.append(motif)
        found_array.append(self.find_motifs(data_frame, motif)[0])
        length_array.append(len(motif))
    
    # Create an array with the total number of possible motifs for each motif
    all_motifs = np.zeros(shape=(len(length_array),))
    for x in range(0, len(length_array)):
        all_motifs[x] = self.proteome_length(data_frame) - length_array[x]

    # Convert into a dictionary and then a Pandas dataframe 
    counted_motifs = pd.DataFrame(list(dict(zip(query_array, found_array)).items()), columns=[&#39;Motif&#39;, &#39;Count&#39;])
    counted_motifs[&#39;Count&#39;][1:].astype(int)  # make the Count column integer format         
    counted_motifs[&#39;Fraction&#39;] = np.asarray(found_array)/np.asarray(all_motifs)  # add a column Fraction for the observed # motifs divided by the total # motifs

    return counted_motifs</code></pre>
</details>
</dd>
<dt id="test.Proteome.expected_motifs"><code class="name flex">
<span>def <span class="ident">expected_motifs</span></span>(<span>self, query_proteome, query_motif)</span>
</code></dt>
<dd>
<div class="desc"><p>Calculate the expected number of a queried motif based on amino acid frequencies in the proteome.</p>
<h2 id="parameters">Parameters</h2>
<p><code>query_proteome</code> : Pandas dataframe
Contains the proteome of interest.</p>
<p><code>query_motif</code> : str or list
Contains the motif(s) of interest.</p>
<h2 id="returns">Returns</h2>
<p><code>final_dataframe</code> :
Pandas dataframe
Expected fraction and count of a given motif.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def expected_motifs(self, query_proteome, query_motif):
    &#34;&#34;&#34; 
    Calculate the expected number of a queried motif based on amino acid frequencies in the proteome.
        
    Parameters
    ----------              
    `query_proteome` : Pandas dataframe
        Contains the proteome of interest.
    
    `query_motif` : str or list
        Contains the motif(s) of interest.
    
    Returns
    -------
    `final_dataframe` :  Pandas dataframe
        Expected fraction and count of a given motif. 
    &#34;&#34;&#34;   
    # Get fractional values of the amino acids within the proteome
    fractional_AA = self.proteome_amino_acid_fractions(query_proteome)
    
    # Get length of the queried proteome 
    length = self.proteome_length(query_proteome)
    
    # Create a dictionary with amino acid-to-number mapping
    numbers = {a: b for b, a in dict(enumerate(self.amino_acids)).items()}
    
    # Check if query_motif is a list of strings
    if isinstance(query_motif, list) == True:
        probabilities = np.zeros(shape=(len(query_motif)))  # Create a numpy array to store probabilities for each motif
        expected = np.zeros(shape=(len(query_motif)))  # Create a numpy array to store expected total for each motif
        observed = np.zeros(shape=(len(query_motif)))  # Create a numpy array to store observed total for each motif
        for i, motif in enumerate(query_motif):
            for res_num in range(0, len(motif)):
                if res_num == 0:
                    probabilities[i] = fractional_AA.at[numbers[motif[res_num]], &#39;Fraction&#39;]  # get the value at the specified position
                else:
                    probabilities[i] *= fractional_AA.at[numbers[motif[res_num]], &#39;Fraction&#39;] # multiply by the value at the specified position
            expected[i] = int(probabilities[i] * (length - len(motif)-1))  # multiply the probability and the total number of motifs
            observed[i] = int(self.find_motifs(query_proteome, motif)[0])
            
        # Combine into final Pandas dataframe    
        final_dataframe = pd.DataFrame({&#39;Motif&#39;:query_motif, &#39;Probability&#39;:probabilities, &#39;Observed&#39;:observed, &#39;Expected&#39;:expected})
    
    
    # Check if query_motif is a string
    elif isinstance(query_motif, str) == True:
        probabilities = np.zeros(shape=(1,))  # Create a numpy array to store probabilities for each motif
        expected = np.zeros(shape=(1,))  # Create a numpy array to store expected total for each motif
        observed = np.zeros(shape=(1,))  # Create a numpy array to store observed total for each motif
        for res_num in range(0, len(query_motif)):
            if res_num == 0:
                probabilities[0] = fractional_AA.at[numbers[query_motif[res_num]], &#39;Fraction&#39;]  # get the value at the specified position
            else:
                probabilities[0] *= fractional_AA.at[numbers[query_motif[res_num]], &#39;Fraction&#39;] # multiply by the value at the specified position
            expected[0] = int(probabilities[0] * (length - len(query_motif)-1))  # multiply the probability and the total number of motifs
            observed[0] = int(self.find_motifs(query_proteome, query_motif)[0])
            
        # Combine into final Pandas dataframe    
        final_dataframe = pd.DataFrame({&#39;Motif&#39;:query_motif, &#39;Probability&#39;:probabilities, &#39;Observed&#39;:observed, &#39;Expected&#39;:expected})
        
    final_dataframe[&#39;Observed&#39;][1:].astype(int)  # make the observed column integer format 
    final_dataframe[&#39;Expected&#39;][1:].astype(int)  # make the expected column integer format
    
    return final_dataframe</code></pre>
</details>
</dd>
<dt id="test.Proteome.find_motifs"><code class="name flex">
<span>def <span class="ident">find_motifs</span></span>(<span>self, data_frame, query)</span>
</code></dt>
<dd>
<div class="desc"><p>Search a proteome for a specific amino acid or motif.</p>
<h2 id="parameters">Parameters</h2>
<p><code>data_frame</code> : Pandas dataframe, contains the proteome of interest</p>
<p><code>query</code> :
str, the amino acid or motif to be searched</p>
<h2 id="returns">Returns</h2>
<p><code>sum</code> :
int, total number of hits for the queried residue/motif</p>
<p><code>fraction</code>:
float, sum divided by the total number of residues or motifs of the same size</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def find_motifs(self, data_frame, query):
    &#34;&#34;&#34; 
    Search a proteome for a specific amino acid or motif.
        
    Parameters
    ----------              
    `data_frame` : Pandas dataframe, contains the proteome of interest
    
    `query` :  str, the amino acid or motif to be searched
    
    Returns
    -------
    `sum` :  int, total number of hits for the queried residue/motif
    
    `fraction`:  float, sum divided by the total number of residues or motifs of the same size
    &#34;&#34;&#34;  
    # Extract sequences from the pandas dataframe
    seqs = data_frame[&#39;Sequence&#39;]
            
    # Calculate the total number of residues in the queried proteome
    total_residues = self.proteome_length(data_frame)
    
    # Initialize an array
    hits_array = []
    
    # Loop over each sequence and count the number of times the queried motif shows up
    for protein in seqs:
        hits_array.append(protein.count(query))

    # total number of hits for the queried residue/motif, check if any hits are found
    sum = np.sum(np.asarray(hits_array))  
    if sum &lt; 1:
        fraction = 0
    else:
        fraction = sum/(total_residues - len(query)+1)  # total number of hits divided by the total number of residues or motifs of the same size
                    
    return sum, fraction       </code></pre>
</details>
</dd>
<dt id="test.Proteome.find_proteins"><code class="name flex">
<span>def <span class="ident">find_proteins</span></span>(<span>self, data_frame, query, proteome_type)</span>
</code></dt>
<dd>
<div class="desc"><p>Search for proteins in a proteome that contain the queried motif.</p>
<h2 id="parameters">Parameters</h2>
<p><code>data_frame</code> : Pandas dataframe, contains the proteome of interest</p>
<p><code>query</code> :
str or list of strings, the amino acid or motifs to be searched (e.g. 'IPV' or ['IPI', 'IPV', 'VPI', 'VPV'])</p>
<p><code>proteome_type</code> : str, either 'FASTA' or 'UniProt' depending on the analyzed proteome</p>
<h2 id="returns">Returns</h2>
<p><code>found_proteins</code> :
list, array of UniProt/FASTA IDs that contain the queried residue/motif</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def find_proteins(self, data_frame, query, proteome_type):
    &#34;&#34;&#34; 
    Search for proteins in a proteome that contain the queried motif.
        
    Parameters
    ----------              
    `data_frame` : Pandas dataframe, contains the proteome of interest
    
    `query` :  str or list of strings, the amino acid or motifs to be searched (e.g. &#39;IPV&#39; or [&#39;IPI&#39;, &#39;IPV&#39;, &#39;VPI&#39;, &#39;VPV&#39;])
    
    `proteome_type` : str, either &#39;FASTA&#39; or &#39;UniProt&#39; depending on the analyzed proteome
    
    Returns
    -------
    `found_proteins` :  list, array of UniProt/FASTA IDs that contain the queried residue/motif
    &#34;&#34;&#34;  
    # First check which type of proteome/Pandas dataframe is being queried
    if proteome_type.upper() == &#39;FASTA&#39;:
        id = &#39;ID&#39;
    elif proteome_type.upper() == &#39;UNIPROT&#39;:
        id = &#39;Entry&#39;
    else:
        print(&#39;ERROR! The proteome_type %s is not recognized in the function &#34;find_proteins&#34; !&#39; % proteome_type)
        print(&#39;Please enter &#34;fasta&#34; or &#34;uniprot&#34;, or modify the function to accommodate other types&#39;)
        print(&#39;Exiting now.........................................................................&#39;)
        sys.exit()

    # Initialize an empty array
    found_proteins = []
    
    # Check if queried motif (i.e. query) is a list
    if isinstance(query, list) == True:
        for i in range(0, len(data_frame[&#39;Sequence&#39;])):
            for j, motif in enumerate(query):
                if motif in data_frame[&#39;Sequence&#39;][i]:
                    found_proteins.append(data_frame[id][i])
          
    # If not list, check if query is a string
    elif isinstance(query, str) == True:
        for i in range(0, len(data_frame[&#39;Sequence&#39;])):
            if query in data_frame[&#39;Sequence&#39;][i]:
                found_proteins.append(data_frame[id][i])

    else:
        print(&#39;In the function &#34;find_proteins&#34;, the &#34;query&#34; must be either a list of strings or a single string....!&#39;)
        print(&#39;Please check what you entered!&#39;)
        print(&#39;Exiting now...................&#39;)
        sys.exit()
    
    # Remove duplicates from list
    found_proteins = list(dict.fromkeys(found_proteins))
    
    return found_proteins</code></pre>
</details>
</dd>
<dt id="test.Proteome.iterate_motif"><code class="name flex">
<span>def <span class="ident">iterate_motif</span></span>(<span>self, query)</span>
</code></dt>
<dd>
<div class="desc"><p>Generate a list of all possible motifs that are consistent with an input motif </p>
<p>Note the following usage specifications:</p>
<p>(1) the "_" symbol separates positions in the motif</p>
<p>(2) "X" designates any residue</p>
<p>(3) multiple residues indicate "or" (e.g. IV_X means Ile or Val in the 1st position followed by any residue)</p>
<p>Example:
MILV_X_DE means Met, Ile, Leu, or Val in the 1st position followed by any residue in the 2nd position followed by Asp or Glu in the 3rd position.</p>
<h2 id="parameters">Parameters</h2>
<p><code>query</code> : str, the amino acid motif to be queried
use "_" to separate positions in the motif, e.g. I/V-X-I/V would be IV_X_IV</p>
<h2 id="returns">Returns</h2>
<p><code>all_possibilities</code> :
list, all variations of the inputted motif (e.g. IV_X_IV yields [IAI, ICI, IDI, &hellip; , VVV, VWV, VYV]</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def iterate_motif(self, query):
    &#34;&#34;&#34; 
    Generate a list of all possible motifs that are consistent with an input motif 
    
    Note the following usage specifications:
    
    (1) the &#34;_&#34; symbol separates positions in the motif
    
    (2) &#34;X&#34; designates any residue
    
    (3) multiple residues indicate &#34;or&#34; (e.g. IV_X means Ile or Val in the 1st position followed by any residue)
    
    Example:  MILV_X_DE means Met, Ile, Leu, or Val in the 1st position followed by any residue in the 2nd position followed by Asp or Glu in the 3rd position.
    
    Parameters
    ----------              
    `query` : str, the amino acid motif to be queried 
            use &#34;_&#34; to separate positions in the motif, e.g. I/V-X-I/V would be IV_X_IV
    
    Returns
    -------
    `all_possibilities` :  list, all variations of the inputted motif (e.g. IV_X_IV yields [IAI, ICI, IDI, ... , VVV, VWV, VYV]
    &#34;&#34;&#34; 
    # Split the input query into its corresponding residue positions
    listq = [item for item in query.split(&#34;_&#34;)]

    # Generator an iterable list of motifs  
    positions = [[] for q in listq]
    for i in range(len(listq)):
        if len(listq[i]) == 1:
            if listq[i] == &#34;X&#34;:
                for aa in self.amino_acids:
                    positions[i].append(aa)
            else:
                positions[i].append(listq[i])
        elif len(listq[i]) &gt; 1:
            for j in listq[i]:
                positions[i].append(j)
        else:
            print(&#34;WARNING! iterate_motif does not recognize your input&#34;)
            print(&#39;exiting now....&#39;)
            sys.exit()
    
    # Use itertools to generate the final list 
    all_possibilities = [&#39;&#39;.join(x) for x in list(itertools.product(*positions))]

    return all_possibilities</code></pre>
</details>
</dd>
<dt id="test.Proteome.load_fasta"><code class="name flex">
<span>def <span class="ident">load_fasta</span></span>(<span>self, fasta)</span>
</code></dt>
<dd>
<div class="desc"><p>Read a FASTA file with identifiers and sequences.</p>
<p>Note that the FASTA file can be single- or multi-line.</p>
<h2 id="parameters">Parameters</h2>
<p><code>fasta</code> : file in FASTA format with identifiers (e.g. &gt;) and sequences </p>
<h2 id="returns">Returns</h2>
<p><code>fasta_proteome</code>, a Pandas dataframe that contains the FASTA IDs, sequences, and length of sequences</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def load_fasta(self, fasta):
    &#34;&#34;&#34; 
    Read a FASTA file with identifiers and sequences.
    
    Note that the FASTA file can be single- or multi-line.
        
    Parameters
    ----------              
    `fasta` : file in FASTA format with identifiers (e.g. &gt;) and sequences 
    
    Returns
    -------
    `fasta_proteome`, a Pandas dataframe that contains the FASTA IDs, sequences, and length of sequences
    &#34;&#34;&#34;
    # Load the fasta file
    self.seq_list = SeqIO.index(fasta, &#34;fasta&#34;)  # makes dictionary that contains Seq objects

    # Create arrays to store ID and sequence information
    ids = []  ; seqs = []
    for key, value in self.seq_list.items():
        ids.append(key)
        seqs.append(str(value.seq)) # convert the Seq object to a string

    # Create a dictionary for subsequent conversion into a dataframe
    self.dictionary = {&#39;ID&#39;:ids, &#39;Sequence&#39;:seqs} 

    # Load the dictionary into a dataframe
    self.fasta_proteome = pd.DataFrame(self.dictionary)  
    self.fasta_proteome[&#39;Length&#39;] = self.fasta_proteome[&#39;Sequence&#39;].str.len()  # create column in the dataframe with sequence lengths
    self.fasta_proteome[&#39;Length&#39;][1:].astype(int)   # convert the length values to integers

    return self.fasta_proteome</code></pre>
</details>
</dd>
<dt id="test.Proteome.load_uniprot"><code class="name flex">
<span>def <span class="ident">load_uniprot</span></span>(<span>self, uniprot)</span>
</code></dt>
<dd>
<div class="desc"><p>Read a UniProt proteome file.</p>
<p>Note that the input proteome file should contain the following column order:
(1) Entry (2) Entry name (3) Status (4) Protein names (5) Gene names (6) Organism (7) Length (8) Sequence </p>
<h2 id="parameters">Parameters</h2>
<p><code>uniprot</code> : file downloaded from UniProt containing a proteome of interest</p>
<h2 id="returns">Returns</h2>
<p><code>proteome_dataframe</code>, a Pandas dataframe that contains the FASTA IDs, sequences, and length of sequences</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def load_uniprot(self, uniprot):
    &#34;&#34;&#34; 
    Read a UniProt proteome file.
    
    Note that the input proteome file should contain the following column order:
    (1) Entry (2) Entry name (3) Status (4) Protein names (5) Gene names (6) Organism (7) Length (8) Sequence 
        
    Parameters
    ----------              
    `uniprot` : file downloaded from UniProt containing a proteome of interest

    
    Returns
    -------
    `proteome_dataframe`, a Pandas dataframe that contains the FASTA IDs, sequences, and length of sequences 
    &#34;&#34;&#34;        
    # Load the dataset with column headers defined
    self.uniprot_proteome = pd.read_csv(uniprot, sep=&#39;\t&#39;, names=[&#39;Entry&#39;, &#39;Entry name&#39;, &#39;Status&#39;, &#39;Protein names&#39;, &#39;Gene names&#39;, &#39;Organism&#39;, &#39;Length&#39;, &#39;Sequence&#39;])
    
    # Convert protein length values to integers
    self.uniprot_proteome[&#39;Length&#39;][1:] =  self.uniprot_proteome[&#39;Length&#39;][1:].astype(int)  
    
    return self.uniprot_proteome</code></pre>
</details>
</dd>
<dt id="test.Proteome.proteome_amino_acid_fractions"><code class="name flex">
<span>def <span class="ident">proteome_amino_acid_fractions</span></span>(<span>self, query_proteome)</span>
</code></dt>
<dd>
<div class="desc"><p>Calculate the fractional amino acid composition of a proteome.</p>
<h2 id="parameters">Parameters</h2>
<p><code>query_proteome</code> : Pandas dataframe, contains the proteome of interest</p>
<h2 id="returns">Returns</h2>
<p><code>AA_fractions</code> :
Pandas dataframe, the fraction of each amino acid</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def proteome_amino_acid_fractions(self, query_proteome):
    &#34;&#34;&#34; 
    Calculate the fractional amino acid composition of a proteome.
        
    Parameters
    ----------              
    `query_proteome` : Pandas dataframe, contains the proteome of interest
    
    Returns
    -------
    `AA_fractions` :  Pandas dataframe, the fraction of each amino acid
    &#34;&#34;&#34;   
    # Amino acids 
    self.amino_acids = [&#39;A&#39;, &#39;C&#39;, &#39;D&#39;, &#39;E&#39;, &#39;F&#39;, &#39;G&#39;, &#39;H&#39;, &#39;I&#39;, &#39;K&#39;, &#39;L&#39;, &#39;M&#39;, &#39;N&#39;, &#39;P&#39;, &#39;Q&#39;, &#39;R&#39;, &#39;S&#39;, &#39;T&#39;, &#39;V&#39;, &#39;W&#39;, &#39;Y&#39;]
    
    # Initialize an empty array to store the fractional amino acid values
    fractions = []

    # Compute the fractional amino acid composition of the proteome
    for aa in self.amino_acids:
        fractions.append(self.find_motifs(query_proteome, aa)[1])
        
    # Convert into a dictionary and then a Pandas dataframe 
    AA_fractions = pd.DataFrame(list(dict(zip(self.amino_acids, fractions)).items()), columns=[&#39;AA&#39;, &#39;Fraction&#39;])
    AA_fractions[&#39;Fraction&#39;][1:].astype(float) # convert the fractions to floats
    
    return AA_fractions  </code></pre>
</details>
</dd>
<dt id="test.Proteome.proteome_length"><code class="name flex">
<span>def <span class="ident">proteome_length</span></span>(<span>self, query_proteome)</span>
</code></dt>
<dd>
<div class="desc"><p>Calculate the total number of amino acids in a proteome.</p>
<h2 id="parameters">Parameters</h2>
<p><code>query_proteome</code> : Pandas dataframe, contains the proteome of interest</p>
<h2 id="returns">Returns</h2>
<p><code>total_length</code> :
int, total number of residues in proteome_dataframe</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def proteome_length(self, query_proteome):
    &#34;&#34;&#34; 
    Calculate the total number of amino acids in a proteome.
        
    Parameters
    ----------              
    `query_proteome` : Pandas dataframe, contains the proteome of interest
    
    Returns
    -------
    `total_length` :  int, total number of residues in proteome_dataframe
    &#34;&#34;&#34;   
    # Sum up the total number of residues
    total_length = query_proteome[&#39;Length&#39;][1:].sum(axis=0, skipna=True)
    
    return total_length</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="test.Proteome" href="#test.Proteome">Proteome</a></code></h4>
<ul class="">
<li><code><a title="test.Proteome.count_multiple_motifs" href="#test.Proteome.count_multiple_motifs">count_multiple_motifs</a></code></li>
<li><code><a title="test.Proteome.expected_motifs" href="#test.Proteome.expected_motifs">expected_motifs</a></code></li>
<li><code><a title="test.Proteome.find_motifs" href="#test.Proteome.find_motifs">find_motifs</a></code></li>
<li><code><a title="test.Proteome.find_proteins" href="#test.Proteome.find_proteins">find_proteins</a></code></li>
<li><code><a title="test.Proteome.iterate_motif" href="#test.Proteome.iterate_motif">iterate_motif</a></code></li>
<li><code><a title="test.Proteome.load_fasta" href="#test.Proteome.load_fasta">load_fasta</a></code></li>
<li><code><a title="test.Proteome.load_uniprot" href="#test.Proteome.load_uniprot">load_uniprot</a></code></li>
<li><code><a title="test.Proteome.proteome_amino_acid_fractions" href="#test.Proteome.proteome_amino_acid_fractions">proteome_amino_acid_fractions</a></code></li>
<li><code><a title="test.Proteome.proteome_length" href="#test.Proteome.proteome_length">proteome_length</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.8.1</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>
